lib/sample_rate_converter/sample_rate_converter.c

```diff
@@ -0,0 +1,407 @@
+
+#ifdef CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_16
+#define SAMPLE_RATE_CONVERTER_INTERNAL_INPUT_BUF_SIZE                                              \
+	INTERNAL_INPUT_BUF_NUMBER_SAMPLES * sizeof(uint16_t)
+#define SAMLPE_RATE_CONVERTER_INTERNAL_OUTPUT_BUF_SIZE                                             \
+	(CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE * sizeof(uint16_t))
+#elif CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_32
+#define SAMPLE_RATE_CONVERTER_INTERNAL_INPUT_BUF_SIZE                                              \
+	INTERNAL_INPUT_BUF_NUMBER_SAMPLES * sizeof(uint32_t)
+#define SAMLPE_RATE_CONVERTER_INTERNAL_OUTPUT_BUF_SIZE                                             \
+	(CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE * sizeof(uint32_t))
+#endif
+
+static int validate_sample_rates(uint32_t input_sample_rate, uint32_t output_sample_rate)
+{
+	if (input_sample_rate > output_sample_rate) {
+		if (input_sample_rate != 48000) {
+			LOG_ERR("Invalid input sample rate for downsampling %d", input_sample_rate);
+			return -EINVAL;
+		}
+
+		if ((output_sample_rate != 24000) && (output_sample_rate != 16000)) {
+			LOG_ERR("Invalid output sample rate for downsampling %d",
+				output_sample_rate);
+			return -EINVAL;
+		}
+	} else if (input_sample_rate < output_sample_rate) {
+		if (output_sample_rate != 48000) {
+			LOG_ERR("Invalid output sample rate for upsampling: %d",
+				output_sample_rate);
+			return -EINVAL;
+		}
+
+		if ((input_sample_rate != 24000) && (input_sample_rate != 16000)) {
+			LOG_ERR("Invalid input sample rate for upsampling: %d", input_sample_rate);
+			return -EINVAL;
+		}
+	} else {
+		LOG_ERR("Input and out samplerates are the same");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+/**
+ * When upsampling from 16kHz to 48kHz the input and output bytes must be buffered. This is to
+ * fulfill the requirement for the filter that the number of input bytes must be divisible by the
+ * conversion factor. This function returns true when this is the case.
+ */
+static inline bool conversion_needs_buffering(struct sample_rate_converter_ctx *ctx,
+					      uint8_t conversion_ratio)
+{
+	return ((ctx->conversion_direction == CONVERSION_DIR_UP) && (conversion_ratio == 3));
+}
+
+static inline uint8_t calculate_conversion_ratio(uint32_t input_sample_rate,
+						 uint32_t output_sample_rate)
+{
+	if (input_sample_rate > output_sample_rate) {
+		return input_sample_rate / output_sample_rate;
+	} else {
+		return output_sample_rate / input_sample_rate;
+	}
+}
+
+/**
+ * @brief Reconfigures the sample rate converter context.
+ *
+ * @details Validates and sets all sample rate conversion parameters for the context. If buffering
+ *	    is needed for the conversion, the input buffer will be padded with two samples to
+ *	    ensure there will always be enough samples for a valid conversion.
+ *
+ * @param[in,out]	ctx			Pointer to the sample rate conversion context.
+ * @param[in]		input_sample_rate	Sample rate of the input samples.
```

Comment:
Usually, we go from generic to specific in names. Suggest `sample_rate_input `and `sample_rate_output`